packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,224 @@
+    this.root = {
+      segment: rootSegment,
+      insertionIndex: -1,
+    };
+  }
+
+  /**
+   * Inserts a route into the route tree.
+   * @param route - The route to insert.
+   * @param metadata - The metadata associated with the route.
+   */
+  insert(route: string, metadata: Omit<RouteTreeNodeMetadata, 'route'>): void {
+    let node = this.root;
+    const normalizedRoute = stripTrailingSlash(route);
+    const segments = normalizedRoute.split('/');
+
+    for (const segment of segments) {
+      // Parameterized route segment, replace it with a wildcard
+      const normalizedSegment = segment[0] === ':' ? '*' : segment;
+      let childNode = node.children?.get(normalizedSegment);
+
+      if (!childNode) {
+        childNode = {
+          segment: normalizedSegment,
+          insertionIndex: -1,
+        };
+        node.children ??= new Map();
+        node.children.set(normalizedSegment, childNode);
+      }
+
+      node = childNode;
+    }
+
+    // Store the full route at the leaf node
+    node.metadata = {
+      route: normalizedRoute,
+      ...metadata,
+    };
+    node.insertionIndex = this.insertionIndexCounter++;
+  }
+
+  /**
+   * Matches a route against the route tree, returning the best match
+   * based on the lowest insertion index.
+   * @param route - The route to match.
+   * @returns The best matching route or null if no match is found.
+   */
+  match(route: string): RouteTreeNodeMetadata | null {
+    let bestMatchMetadata: RouteTreeNodeMetadata | null = null;
+    let lowestIndex = Infinity;
+
+    /**
+     * Recursively traverses the route tree to find the best match.
+     */
+    const traverse = (currentNode: RouteTreeNode, remainingSegments: string[]): void => {
+      const { insertionIndex, metadata, children } = currentNode;
+      if (!remainingSegments.length) {
+        // Reached a leaf node, check if it's a better match
+        if (metadata && insertionIndex < lowestIndex) {
+          bestMatchMetadata = metadata;
+          lowestIndex = insertionIndex;
+        }
+
+        return;
+      }
+
+      if (!children) {
+        return;
+      }
+
+      const [segment, ...restSegments] = remainingSegments;
+
+      // Try to match the exact segment
+      let childNode = children.get(segment);
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+
+      // Try to match with a wildcard
+      childNode = children.get('*');
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+    };
+
+    const segments = stripTrailingSlash(route).split('/');
+    traverse(this.root, segments);
+
+    return bestMatchMetadata;
+  }
+
+  /**
+   * Converts the route tree into a nested object representation.
+   * @returns A nested object where each key is a route segment and each value is an object containing
+   *          metadata and possibly further nested route segments.
+   */
+  toObject(): RouteNodeObjectRepresentation {
```

Comment:
Question: What is the use case for `toObject` and `fromObject`? They seem to only be used in tests.

Reply:
These will eventually be used to create, store, and restore the routes from the manifest. This approach is designed to avoid building the routes on the first request, which can be critical for performance in edge workers or serverless deployments. Without this, the application would need to be bootstrapped and rendered multiple times for each request, impacting performance.

Reply:
We need to make this serializable, as it will be stored in the manifest, which is generated during the build process. The manifest cannot include non-native import symbols.

Initializing a tree from a non-serializable format would add a bit more complexity.